Big timer subsystem rework

andyross opened the pull request:
[Continuing the work in the previous refactoring patch series (which appears at the bottom of the stack here too -- we can merge either one), this takes advantage of further API simplification to rework and shrink the timeslicing and timeout backends.

The code to focus on here is the new timeout.c file that largely replaces sys_clock.c (modulo some bits that were reorganized) and the timeslicing rework in sched.c.

Note that all use of the legacy timer driver API bits have been removed.  The existing drivers still retain all the complexity, but are now susceptible to rewrites themselves which would shrink them enormously.  That's the next step.

Commit:
sys_clock.h: Remove asm guards

This header doesn't get included in assembly context, nor does it
provide any asm-usable macros.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel.h: Header hygine, move clock/timer handling

The kernel.h file had a bunch of internal APIs for timeout/clock
handling mixed in.  Move these to sys_clock.h, which it always
included (in a weird location, so move THAT to kernel_includes.h with
everything else).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
sys_clock.h: Make "global variable" APIs into proper functions

The existing API defined sys_clock_{hw_cycles,ticks}_per_sec as simple
"variables" to be shared, except that they were only real storage in
certain modes (the HPET driver, basically) and everywhere else they
were a build constant.

Properly, these should be an API defined by the timer driver (who
controls those rates) and consumed by the clock subsystem.  So give
them function syntax as a stepping stone to get there.

Note that this also removes the deprecated variable
_sys_clock_us_per_tick rather than give it the same treatment.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
clock: Remove CONFIG_TICKLESS_KERNEL_TIME_UNIT_IN_MICRO_SECS

This was only used in a few places just to indirect the already
perfectly valid SYS_CLOCK_TICKS_PER_SEC value.  There's no reason for
these to ever have been kconfig units, and in fact the distinction
appears to have introduced a hidden/untested bug in the power
subsystem (the two variables were used interchangably, but they were
defined in reciprocal units!).

Just use "ticks" as our time unit pervasively, and clarify the docs to
explain that.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
sys_clock.h: Remove sys_clock_ticks_per_sec()

This just got turned into a function from a "variable" API, but
post-the-most-recent-patch it turns out to be degenerate anyway.
Everyone everywhere should always have been using the kconfig variable
directly, and it was only a weirdness in the tickless API that made it
confusing.  Fix.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
sys_clock: Make sys_clock_hw_cycles_per_tick() a proper API

This was another "global variable" API.  Give it function syntax too.
Also add a warning, because on nRF devices (at least) the cycle clock
runs in kHz and is too slow to give a precise answer here.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
sys_clock: Make clock_always_on true by default

This flag is an indication to the timer driver that the OS doesn't
care about rollover conditions of the tick count while idling, so the
system doesn't need to wake up once per counter flip[1].  Obviously in
that circumstance values returned from k_uptime_get_32() are going to
be wrong, so the implementation had an assert to check for misuse.

But no one understood that from the docs, so the only place these APIs
were used in practice were as "guards" around code that needed to call
k_uptime_get_32(), even though that's 100% wrong per docs!

Clarify the docs.  Remove the incorrect guards.  Change the flag to
initialize to true so that uptime isn't broken-by-default in tickless
mode.  Also move the implemenations of the functions out of the
header, as there's no good reason for these to need to be inlined.

[1] Which can be significant.  A 100MHz ARM using the 24 bit SysTick
    counter rolls over at about 6 Hz, and if it had to come out of
    idle at that rate it would be a significant power issue that would
    swamp the gains from tickless.  Obviously systems with slow
    counters like nRF or 64 bit ones like RISC-V or x86's TSC aren't
    as affected.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
sys_clock: Fix unsafe tick count usage

The system tick count is a 64 bit quantity that gets updated from
interrupt context, meaning that it's dangerously non-atomic and has to
be locked.  The core kernel clock code did this right.

But the value was also exposed to the rest of the universe as a global
variable, and virtually nothing else was doing this correctly.  Even
in the timer ISRs themselves, the interrupts may be themselves
preempted (most of our architectures support nested interrupts) by
code that wants to set timeouts and inspect system uptime.

Define a z_tick_{get,set}() API, eliminate the old variable, and make
sure everyone uses the right mechanism.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
system_timer.h: Remove ASMLANGUAGE guard

This header isn't actually needed in the one assembly context where
it's included.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
sys_clock: Fix up tick announce API

There were three separate "announce ticks" entry points exposed for
use by drivers.  Unify them to just a single z_clock_announce()
function, making the "final" tick announcement the business of the
driver only, not the kernel.

Note the oddness with "_sys_idle_elapsed_ticks": this was a global
variable exposed by the kernel.  But it was never actually used by the
kernel.  It was updated and inspected only within the timer drivers,
and only so that it could be passed back to the kernel as the default
(actually hidden) argument to the announce function.  Break this false
dependency by putting this variable into each timer driver
individually.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/timer: Unify timeout setting APIs

The existing API had two almost identical functions: _set_time() and
_timer_idle_enter().  Both simply instruct the timer driver to set the
next timer interrupt expiration appropriately so that the call to
z_clock_announce() will be made at the requested number of ticks.  On
most/all hardware, these should be implementable identically.

Unfortunately because they are specified differently, existing drivers
have implemented them in parallel.

Specify a new, unified, z_clock_set_timeout().  Document it clearly
for implementors.  And provide a shim layer for legacy drivers that
will continue to use the old functions.

Note that this patch fixes an existing bug found by inspection: the
old call to _set_time() out of z_clock_announce() failed to test for
the "wait forever" case in the situation where clock_always_on is
true, meaning that a system that reached this point and then never set
another timeout would freeze its uptime clock incorrectly.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
include/system_timer.h: Timer API cleanup

Rename timer driver API functions to be consistent.  ADD DOCS TO THE
HEADER so implementations understand what the requirements are.
Remove some unused functions that don't need declarations here.

Also removes the per-platform #if's around the power control callback
in favor of a weak-linked noop function in the driver initialization
(adds a few bytes of code to default platforms -- we'll live, I
think).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
timer: Clean up hairy tickless APIs

The tickless driver had a bunch of "hairy" APIs which forced the timer
drivers to do needless low-level accounting for the benefit of the
kernel, all of which then proceeded to implement them via cut and
paste.  Specifically the "program_time" calls forced the driver to
expose to the kernel exactly when the next interrupt was due and how
much time had elapsed, in a parallel API to the existing "what time is
it" and "announce a tick" interrupts that carry the same information.

Remove these from the kernel, replacing them with synthesized logic
written in terms of the simpler APIs.

In some cases there will be a performance impact due to the use of the
64 bit uptime call, but that will go away soon.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
idle: Fix tickless timeout behavior

If the idle code was detecting that it needed to sleep for less than
CONFIG_SYS_TICKLESS_IDLE_THRESH, then it would never call
z_clock_set_timeout() at all, which means that the system would never
wake up unless it already had a timeout scheduled!  Apparently we
lacked a test case to detect this condition.

Honestly this seems like a crazy feature to me.  There's no benefit in
delivering needless tick announcements.  If the system has the
capacity to enter deeper sleep for long timeouts, that's already
exposed via the PM APIs, the timer subsystem needn't be involved.
But... we actually have a test (tickless_concept) that looks at this,
so support it for now and consider deprecation later.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
idle: Remove needless "expired" logic in sys_power_save_idle()

This code (just refactored as part of the timer API work) turns out to
be needless.  It's trying to detect the case where we're being asked
to idle for zero time, but that's not possible with a properly
functioning timer driver: the call to z_clock_announce() must happen
out of an interrupt, and this is the idle thread, which must sit below
any possible interrupt priority.  The call to z_clock_uptime() must
not ever return "too late" until after the timer interrupt has fired,
at which point we'll be inspecting the next timeout (which itself is
guaranteed to be in the future for the same reason).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/timeoutq: Uninline the timeout methods

There was no good reason to have these rather large functions in a
header.  Put them into sys_clock.c for now, pending rework to the
system.

Now the API is clearly visible in a small header.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sys_clock.c: Add asserts to watch dueling "set time" APIs

The current API has an rather unfortuate collision between two APIs:
z_clock_announce(), which is called out of the timer interrupt to
inform the kernel of time passage (and which is responsible for
invoking timer callbacks), and z_tick_set(), which is ALSO used by the
timer drivers for... confusing and inconsistent purposes.

This is sort of a mess.  The tick_set API needs to go away, but before
that I'm adding some assertions to at least make sure the existing
drivers are using them consistently.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Remove clock_always_on control from k_busy_wait()

This feature was a useless noop based on mistaken API understanding.

The idea seems to have been that k_busy_wait() included guards to
ensure "clock_always_on" was true duing the loop, presumably because
the original author was afraid that "turning the clock off" would
affect the operation of k_cycle_get_32().

Then later someone came around and "optimized" this for Quark SE,
where the cycle counter is the RTC and unrelated to the timer driver
used by the clock_always_on feature.  (Except even there it presumably
should have been done at the SoC level and not just in the C1000
devboard -- note that Arduino 101 never would have gotten this).

But it was all a mistake: "clock_always_on" has nothing to do with
en/disabling the system cycle timer (which never happens when the
system is active, that's a feature of idle), it's a control over the
delivery of timer interrupts.  And needless to say we don't care about
timer interrupts when we're spinning on a cycle counter.

Yank the whole mess.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
sys_clock.h: Remove variance of _TICK_ALIGN with TICKLESS_KERNEL

Not sure why this was here.  The point to this API (which is poorly
explained) is to "round up" requested timeout values to an integer
number of ticks in the future, so the timeouts don't expire too soon.

There's no change of that requirement in tickless mode.  While the
"tick" unit will typicaly be a much smaller time (and thus much less
likely to have this kind of aliasing bug), we STILL don't want early
expiration.

And as with everything else in tickless, changing this breaks no
tests.  So remove it as a needless TICKLESS dependency.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sys_clock: Fix build when !SYS_CLOCK_EXISTS

This got broken.  Add some #ifery to handle the case.  Not clean, will
clean up in a future pass once the API is final.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/timer: Add default z_clock_set_timeout() fallback

Useful for tick-only drivers like Pulpino that don't support this.
Ideally we'd have a header-level interface definition for individual
timer drivers to eliminate the noop function call, but this is clean
for now (even the Pulpino hardware looks like it should support
timeouts just fine, so effort would be better spent there than on a
clean "ticked" interface).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: New timeslicing implementation

Instead of checking every time we hit the low-level context switch
path to see if the new thread has a "partner" with which it needs to
share time, just run the slice timer always and reset it from the
scheduler at the points where it has already decided a switch needs to
happen.  In TICKLESS_KERNEL situations, we pay the cost of extra timer
interrupts at ~10Hz or whatever, which is low (note also that this
kind of regular wakeup architecture is required on SMP anyway so the
scheduler can "notice" threads scheduled by other CPUs).  Advantages:

1. Much simpler logic.  Significantly smaller code.  No variance or
   dependence on tickless modes or timer driver (beyond setting a
   simple timeout).

2. No arch-specific assembly integration with _Swap() needed

3. Better performance on many workloads, as the accounting now happens
   at most once per timer interrupt (~5 Hz) and true rescheduling and
   not on every unrelated context switch and interrupt return.

4. It's SMP-safe.  The previous scheme kept the slice ticks as a
   global variable, which was an unnoticed bug.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/include: Move stubs for timeout functions to their declarations

The timeout_q.h scheme, where it declared real functions, but the
stubs for when there was no clock were in wait_q.h was senselessly
weird.  Put them in the same file.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/timeout: Prepare unification of timeout/thread wait_q fields

The existing timeout API wants to store a wait_q on which the thread
is waiting, but it only uses that value in one spot (and there only as
a boolean flag indicating "this thread is waiting on a wait_q).

As it happens threads can already store their own backpointers to a
wait_q (needed for the SCALABLE scheduler backend), so we should use
that instead.

This patch doesn't actually perform that unification yet.  It
reorgnizes things such that the pended_on field is always set at the
point of timeout interaction, and adds a bunch of asserts to make 100%
sure the logic is correct.  The next patch will modify the API.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/timeout: Remove timeout wait_q field

Per previous patch, this is known to be identical with
thread->pended_on.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/timeout: Eliminate wait_q parameters from API

Now that this is known to be an unused value, remove it from the API.
Note that this caught a few spots where we were passing values (a
non-NULL wait_q with a NULL thread handle) that were always being
ignored before.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/timeout: Refactor API

Add the callback parameter to add_timeout(), and remove the thread
argument.  Now the "low level" timeout API can be expressed without
reference to threads.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
system_timer.h: Change "now" uptime API to be simpler for drivers

The current z_clock_uptime() call (recently renamed from
_get_elapsed_program_time) requires the driver to track a full 64 bit
uptime value in ticks, which is entirely separate from the one the
kernel is already keeping.

Don't do that.  Just ask the drivers to track uptime since the last
call to z_clock_announce(), since that is going to map better to
built-in hardware capability.

Obviously existing drivers already have this feature, so they're
actually getting slightly larger in order to implement the new API in
terms of the old one.  But future drivers will thank us.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: New timeout implementation

Now that the API has been fixed up, replace the existing timeout queue
with a much smaller version.  The basic algorithm is unchanged:
timeouts are stored in a sorted dlist with each node nolding a delta
time from the previous node in the list; the announce call just walks
this list pulling off the heads as needed.  Advantages:

* Properly spinlocked and SMP-aware.  The earlier timer implementation
  relied on only CPU 0 doing timeout work, and on an irq_lock() being
  taken before entry (something that was violated in a few spots).
  Now any CPU can wake up for an event (or all of them) and everything
  works correctly.

* The *_thread_timeout() API is now expressible as a clean wrapping
  (just one liners) around the lower-level interface based on function
  pointer callbacks.  As a result the timeout objects no longer need
  to store backpointers to the thread and wait_q and have shrunk by
  33%.

* MUCH smaller, to the tune of hundreds of lines of code removed.

* Future proof, in that all operations on the queue are now fronted by
  just two entry points (_add_timeout() and z_clock_announce()) which
  can easily be augmented with fancier data structures.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Move timeout_remaining API

_timeout_remaining_get() was a function on a struct _timeout, doing
iteration on the timeout list, but it was defined in timer.c (the
higher level abstraction).

Move it to where it belongs.  Also have it return ticks instead of ms
to conform to scheme in the rest of the timeout API.  And rename it to
a more standard zephyr name.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel: Bump stack size for mem_protect/stackprot

This test needs just a tiny bit of extra stack.  512 bytes isn't
enough on x86 with the most recent set of timer patches.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Rebased to catch up with @ceolin's coding style fixes.

And, I figured I'd see what happens when TICKLESS_KERNEL becomes default.  And everything passes!  So tack on that last patch.  I think everything is good to go here, though surely at least one bug exists on a hardware platform waiting to be discovered.

Work on timer driver rewrites/simplification can happen in future PR's.  The older drivers are working just great when used in the new style.  Likewise there is surely some useful refactoring left to do.  But nothing that can't wait.

carlescufi commented:
Copying additional reviewers due to GitHub reviewer limit. @pizi-nordic @dcpleung

aescolar commented on drivers/timer/legacy_api.h:
```diff
@@ -0,0 +1,69 @@
+ * in terms of the original APIs.  Used by older timer drivers.
+ * Should be replaced.
+ *
+ * Yes, this "header" includes function definitions and must be
```
> Maybe you want to declare the function static to save this note?
>> It's a public function. Basically the legacy_api.h "header" is a stub of C code that the driver code can include to implement the new API in terms of the old one. It's grown another wart or two in the next pull request too.

Yes, I got that :). 
Even though static is normally understood as “private”, I was going for the true meaning of static (scope limited to the translation unit)
(This is a very minor thing anyhow)
So, my point was that, if you add static to ```z_clock_set_timeout  & z_clock_elapsed``` below, then the header can be “safely” included multiple times. Just each time it is included, a new copy of that function is created, without name-clashes with wherever it would be included again.

It does not really matter as there would only be 1 driver, and therefore the header should be included only once.
But it would save you from needing to warn people to not need to include the header more than once.

andyross replied:
Yes, but those APIs are exported functions from the timer driver that need to be called by the kernel.  They must be non-static.  But they're also declared in a "header" due to the crazy shim architecture, thus the warning.

Again, it's all sort of academic anyway.  The drivers need to be reworked to drop the whole scheme and this header will go away.

Commit:
kernel: Checkpatch fixups

I was pretty careful, but these snuck in.  Most of them are due to
overbroad string replacements in comments.  The pull request is very
large, and I'm too lazy to find exactly where to back-merge all of
these.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Heh, looked again this morning.  In fact it was indeed too late yesterday -- making TICKLESS_KERNEL default to =y doesn't actually do anything in the kconfig, because the whole tunable only exists when power maangement is enabled.  So most of the testing was a noop.

Looking more clearly this morning, the older timer drivers (at least hpet and systick) do not, in fact, do the right thing in tickless mode.  They fail to call z_clock_announce() out of the interrupt handler, instead relying on the alternative z_set_tick() API.  So I have more work to do before we're tickless everywhere.

pizi-nordic commented on kernel/timeout.c:
```diff
@@ -0,0 +1,243 @@
+	}
+
+	announce_advance = 0;
+	z_clock_set_timeout(_get_next_timeout_expiry(), false);
```
What if  `_get_next_timeout_expiry()` return 0?

andyross replied:
Perfectly legal.  Informs the driver to deliver another interrupt as soon as possible.

pizi-nordic commented on kernel/timeout.c:
```diff
@@ -0,0 +1,243 @@
+		struct _timeout *to = first();
+
+		ret = to == NULL ? max : adjust_elapsed(to->dticks);
+	}
```
Here you an still have race between `adjust_elapsed()` and the `z_clock_set_timeout()`...
We can run whole `z_clock_set_timeout(_get_next_timeout_expiry(), false);` with interrupts locked, but this will not solve potential tick "skipping" due to CPU speed (however it will make it very unlikely).

Is there a possibility to read "now()" only for newly added timeouts, and process existing ones basing only on the values from `z_clock_announce()` ?

Something like this:
- new_timeout = ms_to_ticks(timeout) + z_clock_elapsed();
- z_clock_set_timeout(new_timeout) /* reference: time of last z_clock_announce() */

andyross replied:
FWIW, that race is inherent: the flow of time doesn't stop.  On hardware that have a minimum-future-interrupt time, that's the responsibility of the driver to make sure that the interrupt still happens even if the kernel asks for something very soon (or already expired).  If you check the timerhacks branch, you'll see the beginnings of a rewrite of the HPET driver that has logic for this.

As for the "now()" being timeout-specific, check the logic in z_clock_announce().  I believe it does what you want.  Timeouts added during processing are done relative to the current head of the timeout queue, which is updated with each item pulled off the list.  So if you have a timeout that adds itself, and have missed a clock interrupt or three, then it should still add back a timeout relative to its scheduled expiration and not whatever the wall clock says.

andyross commented on kernel/sys_clock.c:
```diff
@@ -297,6 +300,8 @@ void z_clock_announce(s32_t ticks)
 {
 	z_last_tick_announced += ticks;
 
+	__ASSERT_NO_MSG(z_last_tick_announced >= tick_count);
```
This was here as a temporary check to make sure I didn't break anything during code evolution.  Note that sys_clock.c gets completely deleted later in the series.

andyross commented on kernel/include/timeout_q.h:
```diff
@@ -41,6 +38,14 @@ int _abort_thread_timeout(struct k_thread *thread);
+#else
+
+/* Stubs when !CONFIG_SYS_CLOCK_EXISTS */
+#define _init_thread_timeout(t) do{}while(0)
```
There was a recent kerfuffle with this vs. MISRA.  Apparently our checking tool regards an empty function as "dead code" (which it is not!), but is OK with preprocessor madness.  So I did it this way out of general pique.  @ceolin can explain more, or if there are other workarounds. :)

andyross commented on kernel/sys_clock.c:
```diff
@@ -494,13 +494,13 @@ s32_t _get_next_timeout_expiry(void)
+		  _timeout_func_t fn, s32_t timeout_in_ticks)
 {
 	__ASSERT(timeout_in_ticks >= 0, "");
+	__ASSERT(fn == timeout->func, "");
```
We do!  :)  I moved the callback to be an argument to the add() routine instead of init().  This assertion was added to be sure I didn't break something.

andyross commented on drivers/timer/legacy_api.h:
```diff
@@ -43,7 +43,11 @@ static u32_t driver_uptime;
+#ifdef TICKLESS_KERNEL
 	return (u32_t)(z_clock_uptime() - driver_uptime);
+#else
+	return 0;
```
It's a definitional thing.  "Kernel time" **means** the number of ticks that have been announced.  Except that when not every tick has been announced, you need to ask the driver for the difference.  Apps that need real time metrics beyond ticks need to be using alternate APIs.  Timekeeping to the kernel means ticks.

Commit:
system_timer.h: Update docs

Clarify behavior of the ticks argument to z_clock_set_timeout() and
add an important note about expected behavior in SMP environments.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/kconfig: Move TICKLESS options out of power management tree

These options are rapidly becoming a default configuration, which is
complicated by having them be hidden inside of a SYS_POWER_MANAGEMENT
variable that has to be enabled first.  Put them at the top level of
the kernel config.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Rebased.  Added two minor commits.

Commit:
kernel/sched: Remove spurious locking

The timeout APIs are properly synchronized now.  This irq_lock() (and
the comment explaining it) isn't needed anymore.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Fix timeslicing predicate

It's possible to interrupt a thread that has already scheduled a
timeout.  Really this is a race against the usage of
_add_thread_timeout() and needs some design work to provide proper
locking (which is a distinct requirement from the scheduler lock and
timeout lock!), as the users of that API are spread around the kernel.
But existing usage always schedules the timeouts first, so this is
safe.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/timeout: Fix synchronization in z_tick_get_32()

The previous comment correctly and carefully explained why the 64 bit
value in curr_tick doesn't require locking when reading only the low
32 bits.

It completely missed the fact that the calculation of elapsed time and
the read of curr_tick ABSOLUTELY DO require locking, because the
former is expressed in terms of the latter.  This was always bug, even
in the old code, but never witnessed because we ran so little software
in tickless mode.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
include/drivers: Fix docs on z_clock_set_timeout()

This wasn't explained correctly.  The tick convention we use here
(owing to the way legacy code was written) is a little weird.  Timeout
delays are passed in a "round down" sense, so that setting a timeout
in "one tick" means that the interrupt will arrive anywhere between
zero and one ticks in the future.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/timeout: Fix announcement tick logic

This was wrong in subtle ways.  In tickless mode it's possible to get
an announcement for multiple ticks at a time and have multiple
callbacks to execute that were technically scheduled at different
times.  We want to fix the current tick at the value represented by
the currently-executing callback's EXPIRATION (even if we missed it!),
so that any new timeouts it sets (c.f. a k_timer period) happen at the
right point, in phase with the expected series.  In single-tick mode
the code ends up the same always, so the bug wasn't visible.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/context: Disable test_kernel_interrupts when tickless

The theory behind this test seems to be that taking an IRQ lock should
prevent the advance of the kernel's tick counter.  That works on
traditional timers only.  In tickless mode the timer hardware/driver
is expected to be able to give us an answer for time independent of
interrupt delivery, so the test fails spuriously.  The "bug" detected
is a feature of tickless!

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/timeout: Fix elapsed logic

When fetching the next timeout to expire, the value is relative to the
last announced tick, so you subtract the timer-provided elapsed time
to get the true delta from "now".  When adding a new timeout, you
*have* a value relative to now, so you compute the delta vs. the last
announced tick by adding the elapsed() time.  Duh.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/context: Skip test_kernel_cpu_idle when tickless

This test was written to assume that on idle the CPU would wake up on
the next tick boundary because of the timer interrupt.  No such
interrupt arrives in tickless mode and it hangs forever.

A more whiteboxy test involving setting a clock timout will have to be
written for this feature if we want to keep it on tickless systems.
Alternatively we could move this test out of tests/kernel/context and
always disable tickless.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

tbursztyka commented on kernel/sys_clock.c:
```diff
@@ -42,7 +42,7 @@ volatile u64_t _sys_clock_tick_count;
  * To save power, this should be turned on only when required.
  */
-int _sys_clock_always_on;
+int _sys_clock_always_on = 1;
```
where/when is _enable_sys_clock() routine being called by default? (or maybe you are tackling this issue later)

andyross replied:
It's happening later.  This flag is becoming a hint managed by the kernel.  The driver is responsible for making the decision about what to do with the hardware.  You can take a look at the new systick driver in my timerhacks branch (https://github.com/andyross/zephyr/commit/9296d740ec70e020839c5a21d78955624efac265) to see how it's expected to look.

Basically the kernel decides whether or not to tell the driver to wait for INT_MAX (i.e. give me a tick, but as far in the future as you can) or K_FOREVER (shut it all down, I don't care).

Commit:
kernel/timeout: Clamp ticks argument to lower bound

Our funny convention holds that passing ticks==1 to _add_timeout()
means "at the next tick".  But that means that 1, 0, and all negative
numbers are expected to behave the same.  In ticked mode, that's fine
because it will, after all, expire at the next tick.

But in tickless, the next announcement may be for several ticks, and
that zero will appear to expire "before" the next tick in the
consumption loop.

Make sure all "next tick" expirations look the same.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Fix timeslice accounting for already-elapsed ticks

In tickless mode, not all elapsed ticks may have been announced yet,
so future z_time_slice() calls will include "extra" ticks that we have
to account for when setting up the slice count.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Add more fixes, almost all of them having to do with tickless enablement.

Folks who haven't reviewed this yet, I'd be appreciative if you could do it soon so we can get this merged early in the cycle to find hardware bugs.

Once it's in, I'll start submitting the new timer drivers (I have ARM SysTick and HPET there now) and turning tickless on by default (which is down to the last few failures -- many of which are turning out to be bad assumptions in test code).

andyross commented:
Oh, and I'm just now realizing that a reinstall of my build machine a while back led to an incorrect .gitconfig and many of my patches are not Author'd and Signed-of-by my personal email and not the correct Intel one.  Will fix.

andyross commented:
Fixed my email addresses and rebased

pizi-nordic commented on tests/kernel/timer/timer_api/src/main.c:
```diff
@@ -226,18 +227,20 @@ void test_timer_periodicity(void)
+		 *  of the expected period (firing time)
 		 */
-		TIMER_ASSERT(WITHIN_ERROR(delta, PERIOD, 1), &timer);
+		TIMER_ASSERT(abs(dticks - _ms_to_ticks(PERIOD)) <= 2, &timer);
```
Is firing before the timeout correct? IMHO We can fire a bit later, but we should detect the situation when timer fired too early.

andyross replied:
It's not that it fires before the timeout, it's that the internal computation of when the timeout is going to expire inherently races against the test code's, so if you compute the timestamp before the k_timer_start() call, you don't know that they aren't on opposite sides of the same tick (and vice versa).  This used to be mostly academic as timeout handling always happened out of callbacks (or due to wakeups from callbacks) that only happened on tick boundaries, and the z_tick_get() value would change only on those same boundaries.  But in tickless mode it's realtime.

That said: I'm finding lots and lots of places in the tests that are bitten by this kind of aliasing (you can see the patches in the timerhacks branch), and intuitively they seem to be more common than they should.  For simple timer test code, it SHOULD be possible to have the tickless ordering be identical to ticked, and I'm worried that the fact I have to change so many tests is really evidence of a bug in the systick driver in tickless mode.

I'm removing this particular patch from this series, basically.

andyross commented:
Fixed the checkpatch issues that snuck in yesterday.  Removed one of the "test precision" patches that was in this series.  It's back in the timerhacks branch instead.  There is a whole series of these, enough that I'm starting to get cold feet that they're really showing a bug in the cycle counting logic in the new systick driver there.  They don't belong in this series anyway.

dcpleung commented on kernel/include/timeout_q.h:
```diff
@@ -10,281 +10,55 @@
-	s32_t *delta = &timeout->delta_ticks_from_prev;
-	struct _timeout *in_q;
+/* Stubs when !CONFIG_SYS_CLOCK_EXISTS */
+#define _init_thread_timeout(t) do {} while (0)
```
FYI, MISRA C wants you to use (false) from stdbool.h instead of (0).

andyross commented on include/sys_clock.h:
```diff
@@ -76,9 +77,80 @@ extern int sys_clock_hw_cycles_per_tick;
+
+#ifdef _NEED_PRECISE_TICK_MS_CONVERSION
+	/* use 64-bit math to keep precision */
+	return (u64_t)ticks * MSEC_PER_SEC / CONFIG_SYS_CLOCK_TICKS_PER_SEC;
```
Probably not, but this is unchanged code. It's just being moved.

andyross commented:
For those following this patch, note that there's yet another evolved PR up at #10556 that includes drivers.

nashif commented on kernel/idle.c:
```diff
@@ -13,22 +13,11 @@
-
-#if defined(CONFIG_TICKLESS_KERNEL)
-#define _must_enter_tickless_idle(ticks) (true)
+#ifdef CONFIG_TICKLESS_IDLE_THRESH
```
This patch series is not changing the behaviour or enabling tickless by default, afaik this will come in the followup series

pfalcon commented:
@andyross, @nashif: That feeling, as if it's Zephyr is not many years in development, but we're starting with an early prototype:

~~~
/home/buildslave/workspace/lite-aeolus-micropython/zephyr/kernel/timeout.c:47:13: error: conflicting types for 'remove'
12:22:03  static void remove(struct _timeout *t)
12:22:03              ^~~~~~
12:22:03 In file included from /home/buildslave/workspace/lite-aeolus-micropython/zephyr/ext/hal/st/stm32cube/stm32f4xx/drivers/include/stm32f4xx_hal_def.h:48:0,
12:22:03                  from /home/buildslave/workspace/lite-aeolus-micropython/zephyr/ext/hal/st/stm32cube/stm32f4xx/drivers/include/stm32f4xx_hal_rcc.h:45,
12:22:03                  from /home/buildslave/workspace/lite-aeolus-micropython/zephyr/ext/hal/st/stm32cube/stm32f4xx/drivers/include/stm32f4xx_hal_conf.h:259,
12:22:03                  from /home/buildslave/workspace/lite-aeolus-micropython/zephyr/ext/hal/st/stm32cube/stm32f4xx/drivers/include/stm32f4xx_hal.h:46,
12:22:03                  from /home/buildslave/workspace/lite-aeolus-micropython/zephyr/ext/hal/st/stm32cube/stm32f4xx/soc/stm32f4xx.h:250,
12:22:03                  from /home/buildslave/workspace/lite-aeolus-micropython/zephyr/soc/arm/st_stm32/stm32f4/soc.h:27,
12:22:03                  from /home/buildslave/workspace/lite-aeolus-micropython/zephyr/include/arch/arm/cortex_m/cmsis.h:21,
12:22:03                  from /home/buildslave/workspace/lite-aeolus-micropython/zephyr/arch/arm/include/cortex_m/stack.h:27,
12:22:03                  from /home/buildslave/workspace/lite-aeolus-micropython/zephyr/arch/arm/include/kernel_arch_data.h:52,
12:22:03                  from /home/buildslave/workspace/lite-aeolus-micropython/zephyr/kernel/include/kernel_structs.h:65,
12:22:03                  from /home/buildslave/workspace/lite-aeolus-micropython/zephyr/kernel/include/ksched.h:10,
12:22:03                  from /home/buildslave/workspace/lite-aeolus-micropython/zephyr/kernel/timeout.c:10:
12:22:03 /home/buildslave/srv/toolchain/zephyr-sdk-0.9.3/sysroots/armv5-zephyr-eabi/usr/include/stdio.h:227:5: note: previous declaration of 'remove' was here
12:22:03  int _EXFUN(remove, (const char *));
12:22:03      ^~~~~~
~~~
